### PR TITLE
[node.js] Created routes for recommendation and getting cumulative audio features of current stitches.

### PR DESCRIPTION
**Pull request by DiegoLandaeta03**
**Backend**

- Created routes for recommendations. 
- Made function for fetching 20 recommendations based on the last 5 songs that the user chose.
- I extracted solely the id's so that they can be used later to fetch for their respective audio features.
- Additionally, made a function that fetches the audio features of all of the songs in the stitch. 

**The return of the recommendation call that returns the Spotify id's for the 20 recommendations:**

Screenshot 2024-07-15 at 5 44 06 PM

**Audio features of 2 songs from the stitch selected:**

Screenshot 2024-07-15 at 5 43 39 PM

### CODE REVIEW

**Comment by alegleason** [comment]
Nice test plan! Thanks for the screenshots.

**Comment by alegleason** [comment]
To get the recommendations you're using the 'Get Recommendations' from the API, right? https://developer.spotify.com/documentation/web-api/reference/get-recommendations

Wouldn't that be a bit redundant with the recommendations built on top of the algorithm proper of the web application? Let's go through this on our 1:1.

File: back-end/routes/recommendations.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+const fetch = require('node-fetch');
+const router = express.Router();
+const prisma = new PrismaClient();
+
+router.use(express.json());
+
+const getAccessToken = async (userId) => {
+    const token = await prisma.token.findFirst({
+        where: {
+            userId: userId
```
Comment (alegleason) [inline back-end/routes/recommendations.js:12]: You can just do `userId` // shorthand for userId: userId

File: back-end/routes/recommendations.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    const stringOfIds = songIds.join(',');
+    try {
+        const response = await fetch(`https://api.spotify.com/v1/audio-features?ids=${encodeURIComponent(stringOfIds)}`, {
+            headers: {
+                'Authorization': `Bearer ${access_token}`
+            }
+        });
+        const audioData = await response.json();
+    } catch (error) {
+        console.error(error);
```
Comment (alegleason) [inline back-end/routes/recommendations.js:32]: Maybe just `return null`?

File: back-end/routes/recommendations.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        orderBy: {
+            createdAt: 'desc'
+        }
+    });
+    return token ? token.token : null;
+};
+
+const fetchStitchAverages = async (songs, access_token) => {
+    const songIds = songs.map(song => (song.uri).split(':').pop());
+    const stringOfIds = songIds.join(',');
```
Comment (alegleason) [inline back-end/routes/recommendations.js:23]: nit

`const commaSeparatedIds = songIds.join(',');`

File: back-end/routes/recommendations.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        });
+        const audioData = await response.json();
+    } catch (error) {
+        console.error(error);
+    }
+};
+
+const fetchRecommendations = async (songs, access_token) => {
+    const lastSongIds = songs.map(song => (song.uri).split(':').pop()).slice(-5);
+    const stringOfIds = lastSongIds.join(',');
```
Comment (alegleason) [inline back-end/routes/recommendations.js:38]: nit

`const commaSeparatedIds = songIds.join(',');`

File: back-end/routes/recommendations.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    try {
+        const response = await fetch(`https://api.spotify.com/v1/recommendations?limit=20&seed_tracks=${stringOfIds}`, {
+            headers: {
+                'Authorization': `Bearer ${access_token}`
+            }
+        });
+        const recommendData = await response.json();
+        const recommendations = recommendData.tracks;
```
Comment (alegleason) [inline back-end/routes/recommendations.js:46]: nit

`const trackRecommendations = recommendData.tracks;`

File: back-end/routes/recommendations.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        const recommendIds = recommendations.map(track => (track.uri).split(':').pop());
```
Comment (alegleason) [inline back-end/routes/recommendations.js:47]: nit

`const songRecommendationsIds = trackRecommendations.map(track => (track.uri).split(':').pop());`

File: back-end/routes/recommendations.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        return recommendIds;
+    } catch (error) {
+        console.error(error);
```
Comment (alegleason) [inline back-end/routes/recommendations.js:50]: No `console.error`s as discussed please, let's handle the errors instead.

https://www.freecodecamp.org/news/effective-error-handling-in-react-applications/#how-to-communicate-errors-to-users

File: back-end/routes/recommendations.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                }
+            }
+        });
+
+        if (!stitch) {
+            return res.status(404).json({ error: 'Stitch not found' });
+        }
+
+        const userId = stitch.userId;
+        const access_token = await getAccessToken(userId);
```
Comment (alegleason) [inline back-end/routes/recommendations.js:75]: Please be consistent between snake case and camel case usage.

File: back-end/routes/recommendations.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+        if (!access_token) {
+            return res.status(401).json({ error: 'Access token not found' });
+        }
+
+        const recommendations = await fetchRecommendations(stitch.songs, access_token);
+        const features = await fetchStitchAverages(stitch.songs, access_token);
```
Comment (alegleason) [inline back-end/routes/recommendations.js:82]: nit

`const tracksRecommendations = await fetchRecommendations(stitch.songs, access_token);`
`const tracksFeatures = await fetchStitchAverages(stitch.songs, access_token);`

File: back-end/routes/recommendations.js (hunk 7 above)
Comment (DiegoLandaeta03) [reply back-end/routes/recommendations.js:50]: > No `console.error`s as discussed please, let's handle the errors instead.
> 
> https://www.freecodecamp.org/news/effective-error-handling-in-react-applications/#how-to-communicate-errors-to-users

Can we discuss this? I don't remember us discussing console.error's. Thank you

Comment (alegleason) [reply back-end/routes/recommendations.js:50]: As discussed on earlier PRs*

**Pull request merged**